datafusion/core/src/dataframe.rs

```diff
@@ -627,22 +627,12 @@ impl DataFrame {
     /// # }
     /// ```
     pub async fn count(self) -> Result<usize> {
-        let rows = self
-            .aggregate(
-                vec![],
-                vec![datafusion_expr::count(Expr::Literal(ScalarValue::Null))],
-            )?
-            .collect()
-            .await?;
-        let len = *rows
-            .first()
-            .and_then(|r| r.columns().first())
-            .and_then(|c| c.as_any().downcast_ref::<Int64Array>())
-            .and_then(|a| a.values().first())
-            .ok_or(DataFusionError::Internal(
-                "Unexpected output when collecting for count()".to_string(),
-            ))? as usize;
-        Ok(len)
+        let mut stream = self.execute_stream().await?;
```

Comment:
I'm not sure if there is a better way to express this, in particular I'm not sure what `select count(*) from table` would compile down to.

Reply:
You could keep the old behaviour and use `datafusion_exr::utils::COUNT_STAR_EXPANSION` instead of the `ScalarValue::Null`, since `count(*)` is expanded to that in SQL planner (related bug https://github.com/apache/arrow-datafusion/pull/5518 prevents directly using `Count(Expr:Wildcard)` at the moment)

The reason I had it that way was to let the count optimize out any column projections to only the minimal needed, which I think theoretically is faster than just executing the plan as is and counting the rows